File: days/day22.swift
#!/usr/bin/env swift
import Foundation


func calc(_ seed: Int) -> Int {
    let m = 16777216, n = 2000
    var x = seed
    for _ in 0..<n {
        x = (x ^ (x << 6)) % m
        x = (x ^ (x >> 5)) % m
        x = (x ^ (x &* 2048)) % m
    }
    return x
}


func part1(_ data: String) -> Int {
    let nums = data.matches(of: #/\d+/#).map { Int($0.0)! }

    let ans = nums.map(calc).reduce(0, +)
    return ans
}


struct V4 : Hashable {
    var a, b, c, d: Int

    static func parse<S>(_ data: S) -> V4 where S:RandomAccessCollection<(Int,Int)> {
        V4(
            a: data[data.index(data.startIndex, offsetBy: 0)].1,
            b: data[data.index(data.startIndex, offsetBy: 1)].1,
            c: data[data.index(data.startIndex, offsetBy: 2)].1,
            d: data[data.index(data.startIndex, offsetBy: 3)].1
        )
    }
}


func calcs(_ seed: Int) -> [(Int,Int)] {
    let m = 16777216, n = 2000
    var s = seed % 10
    var x = seed
    return (0..<n).map { _ in
        x = (x ^ (x << 6)) % m
        x = (x ^ (x >> 5)) % m
        x = (x ^ (x &* 2048)) % m
        let k = x % 10
        let r = k - s
        s = k
        return (k, r)
    }
}


func part2(_ data: String) -> Int {
    let nums = data.matches(of: #/\d+/#).map { Int($0.0)! }

    var cache: [V4:Int] = [:]
    for ps in nums.map(calcs) {
        var seen: Set<V4> = []
        for i in (ps.startIndex ..< ps.endIndex - 3) {
            let k = V4.parse(ps[i ..< i + 4])
            if !seen.contains(k) {
                seen.insert(k)
                cache[k, default: 0] += ps[i + 3].0
            }
        }
    }

    return cache.values.max()!
}


let test1 = """
1
10
100
2024
"""
assert(part1(test1) == 37327623)

let test2 = """
1
2
3
2024
"""
assert(part2(test2) == 23)


let data = try! String(contentsOfFile: "day22.in", encoding: .utf8)
let ans1 = part1(data)
print("part1: \(ans1)")
let ans2 = part2(data)
print("part2: \(ans2)")

File: days/day22.py
#!/usr/bin/env python
from collections import defaultdict


def part1(data, M=16777216, N=2000):
    nums = list(map(int, data.split()))

    def calc(x):
        for _ in range(N):
            x = (x ^ (x << 6)) % M
            x = (x ^ (x >> 5)) % M
            x = (x ^ (x * 2048)) % M
        return x

    ans = sum(map(calc, nums))
    return ans


def part2(data, M=16777216, N=2000):
    nums = list(map(int, data.split()))

    def calc(x):
        s = x % 10
        for _ in range(N):
            x = (x ^ (x << 6)) % M
            x = (x ^ (x >> 5)) % M
            x = (x ^ (x * 2048)) % M
            k = (x % 10)
            yield (k, k - s)
            s = k

    seq = [list(zip(*p)) for p in map(calc, nums)]

    cache = defaultdict(int)
    for ps, ks in seq:
        seen = set()
        for i in range(len(ks)-3):
            k = ks[i:i+4]
            if k not in seen:
                seen.add(k)
                cache[k] += ps[i+3]

    ans = max(cache.values())
    return ans


data = '''
1
10
100
2024
'''
assert part1(data) == 37327623

data = '''
1
2
3
2024
'''
assert part2(data) == 23


data = open('day22.in').read()
ans = part1(data)
print('part1:', ans)
ans = part2(data)
print('part2:', ans)
